### ISSUE TITLE
[Question] Does PBG support weighted edges?

### ISSUE THREAD

**Issue by adityat** [facebookresearch/PyTorch-BigGraph#55]
It doesn't seem clear if the algorithm supports weighted edges. If it does should the edge simply be repeated in the input file multiple times? 
 I apologize if it is somewhere in the documentation and I missed it

**Comment by lerks** [comment]
There's several answers here:
- To be technically pedantic, PBG operates on [_multi_-graphs](https://en.wikipedia.org/wiki/Multigraph), i.e., graphs where edges can be repeated (and self-loops are allowed). So, yes, you can repeat the same edge N times in the input and it will be trained upon N times. What this means is that N times each epoch PBG will calculate the loss induced by that edge and use it in the gradient calculation and descent. This however comes with two caveats. First, the replicas of the edge may end up in different batches thus, within each batch, that edge will be weighed the same as all other edges. Second, this only allows you to "encode" integer weights (which, arguably, can be used to approximate any weight distribution, but at an enormous factor in terms of input size).
- Relation _types_ can be weighted, using the `weight` parameter in the [config](https://torchbiggraph.readthedocs.io/en/latest/configuration_file.html). This value must be a real number that will be used as a factor by which the loss of each edge of that type is multiplied when computing the loss. This is more "coarse" than the above (you cannot change the weight of a single edge) but more space-efficient.
- We are working on a change that will allow to attach arbitrary data to each edge. Such data could be a weight. With this change one could have a loss function class that understands this extra data and uses it to weigh the losses of the individual edges before summing them. We expect to implement this soon.

**Comment by anmol1841** [comment]
Just a quick follow-up. Do we have edge weights implemented in the module by now?
I am currently working on a data with edges around 150-200 million. And edge weights gets important as duplicating the edges would exponentially increase the already large data-set.
Apologies if the same question was answered recently

**Comment by anmol1841** [comment]
Just a quick follow-up. Do we have edge weights implemented in the module by now?
I am currently working on a data with edges around 150-200 million. And edge weights gets important as duplicating the edges would exponentially increase the already large data-set.
Apologies if the same question was answered recently.

**Comment by anmol1841** [comment]
Just a quick follow-up. Do we have edge weights implemented in the module by now?
I am currently working on a data with edges around 150-200 million. And edge weight gets important as duplicating the edges would exponentially increase the already large data-set.
Apologies if the same question was answered recently.
